mm2src/kdf_walletconnect/src/lib.rs

```diff
@@ -581,21 +582,41 @@ impl WalletConnectCtxImpl {
         };
         let request = RequestParams::SessionRequest(request);
         let ttl = request.irn_metadata().ttl;
-        self.publish_request(&active_topic, request).await?;
-
-        if let Ok(Some(resp)) = timeout(Duration::from_secs(ttl), async {
-            self.message_rx.lock().await.next().await
+        let message_id = MessageIdGenerator::new().next();
+        self.publish_request(&active_topic, request, message_id).await?;
+
+        match timeout(Duration::from_secs(ttl), async {
+            // Check if the message exists and matches the expected message ID and
+            // wait till we get a message with expected id or timeout.
+            loop {
+                let next_message = {
+                    let mut lock = self.message_rx.lock().await;
+                    lock.next().await
+                };
+
+                if let Some(Ok(message)) = &next_message {
+                    if message.message_id == message_id {
+                        return next_message;
+                    }
+                }
+            }
```

Comment:
ummmm, i imagined it to be something similar to how we do in electrum (also utilizes message-id/rpc-id).
with this setup we would consume the incoming messages out of the channel and discard them instead of passing them to the entity waiting for them.

what i am thinking of is to have a map of response cache and we ask it if these message we got a response for (that's how electrum is done too and i think other rpcs as well).

Reply:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2223/commits/06e4dc7baeb40c8616cb3397fb3c44b91de191e6